### PR TITLE
Create NetworkChart

### PR DESCRIPTION
**Pull request by zorkmark**
New addon for review.

### CODE REVIEW

**Comment by uxcq** [comment]
@zorkmark Any instructions and pictures of what it  looks like when working?

**Comment by zorkmark** [comment]
From the Shakespeare ged database located at https://sourceforge.net/pr
ojects/godskingsheroes/.  
First time I've tried to respond with an SVG picture so not sure if
this will work.  





The default output will give you something similar to this for an
entire database, unless, it is larger than 1500 persons.  If it is
greater than 1500 persons, it will default to show the largest
connected tree in the database.  You can trim this as desired on the
Trim option tab.

For this example I added dynamic links to the nodes based on the
GrampsID.  These can be links from the database as well.  You should be
able to see these links in a browser pointed at the SVG.  This was my
primary objective for NetworkChart.  To show a connectivity graph that
had links so you could use it as a visual index.

I've highlighted a direct path between Judith Shakesphere and Geofrey
Webb by entering only those two people into the Highlight option tab.
 The NetworkChart addon will do the rest.  That path can also be
isolated and only it shown.  This was my secondary objective for
NetworkChart.  This is also the reason you will need to have 
the ##SELECTION_END##networkx python module installed.

The only issue is the fonts in an SVG file.  If you have the White
Rabbit font installed (https://www.fontsquirrel.com/fonts/white-rabbit)
you will see what I see, if not you will see a default font that may or
may not look good.  You can select whichever font you would like though
when you generate the NetworkChart.  It's just the person viewing must
also have that font installed.  

I am working on a description page for the gramps-project.org wiki as
well. 
 
Cheers,

Mark B.

**Comment by zorkmark** [comment]
The SVG did not show up on github##SELECTION_END##. 

The following won't display perfectly but at least it will convey the
idea until I can build the wiki.

https://drive.google.com/file/d/0B020d5C9svlRVmhhaFphRVRSTWM/view

**Comment by uxcq** [comment]
shakespeare_network

File: NetworkChart/NetworkChart.py [hunk 1]
```diff
@@ -0,0 +1,1189 @@
...
+            i_round_marr = 9999
+        try:
+            i_middle_initials = int(float(self.i_middle_initials))
+        except ValueError:
+            i_middle_initials = 9999
+
+        for entry in self.database.get_person_cursor():
+            handle = entry[0] # individual = entry[1]
+            p_person = self.database.get_person_from_handle(handle)
+            person_gref = str(p_person.get_gramps_id())
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: The gramps_id is already a str.

File: NetworkChart/NetworkChart.py [hunk 2]
```diff
@@ -0,0 +1,1189 @@
...
+            p_privacy = p_person.get_privacy()
+            if p_person.urls:
+                p_url = p_person.urls[0].to_struct()['path']
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `p_person.urls[0].get_path()`.

File: NetworkChart/NetworkChart.py [hunk 3]
```diff
@@ -0,0 +1,1189 @@
...
+                p_url = ""
+
+            name = self._name_display.display(p_person)
+
+            bday_index = p_person.birth_ref_index
+            dday_index = p_person.death_ref_index
+
+            if bday_index > -1:
+                bday_handle = p_person.get_event_ref_list()[bday_index].to_struct()['ref']
+                h_event = self.database.get_from_name_and_handle("Event", bday_handle)
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `self.database.get_event_from_handle(bday_handle)`.

File: NetworkChart/NetworkChart.py [hunk 4]
```diff
@@ -0,0 +1,1189 @@
...
+            else:
+                p_url = ""
+
+            name = self._name_display.display(p_person)
+
+            bday_index = p_person.birth_ref_index
+            dday_index = p_person.death_ref_index
+
+            if bday_index > -1:
+                bday_handle = p_person.get_event_ref_list()[bday_index].to_struct()['ref']
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `p_person.get_event_ref_list()[bday_index].ref`.

File: NetworkChart/NetworkChart.py [hunk 5]
```diff
@@ -0,0 +1,1189 @@
...
+                                        name = name.replace(i.strip(), i[0]+'.', 1)
+                    except NameError:
+                        bday = _('b.unk')
+                if bday == _('b.0000-00-00'):
+                    bday = _('b.unk')
+            else:
+                bday = _('b.unk')
+
+            if dday_index > -1:
+                dday_handle = p_person.get_event_ref_list()[dday_index].to_struct()['ref']
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `p_person.get_event_ref_list()[dday_index].ref`.

File: NetworkChart/NetworkChart.py [hunk 6]
```diff
@@ -0,0 +1,1189 @@
...
+                h_event = self.database.get_from_name_and_handle("Event", dday_handle)
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `self.database.get_event_from_handle(dday_handle)`.

File: NetworkChart/NetworkChart.py [hunk 7]
```diff
@@ -0,0 +1,1189 @@
...
+                    dday = _('d.unk')
+            else:
+                dday = _('d.unk')
+
+            gender = const_gender[p_person.get_gender()]
+            person = person + [[person_gref, name, bday, dday, gender, p_privacy, p_url]]
+
+        for handle, family in self.database.get_family_cursor():
+            family_gref = str(family[1])
+            f_family = self.database.get_from_name_and_handle("Family", handle)
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `self.database.get_family_from_handle(handle)`.

File: NetworkChart/NetworkChart.py [hunk 8]
```diff
@@ -0,0 +1,1189 @@
...
+                    self.database.get_person_from_handle(
+                        f_family.get_mother_handle()).get_gramps_id())
+            else:
+                mother_gref = ''
+
+            # Get edge_marriage
+            h_events = f_family.get_event_list() #get_event_ref_list()
+            if len(h_events) > 0:
+                for handle in h_events:
+                    i_event = self.database.get_from_name_and_handle(
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `self.database.get_event_from_handle(handle)`.

File: NetworkChart/NetworkChart.py [hunk 9]
```diff
@@ -0,0 +1,1189 @@
...
+
+            name = self._name_display.display(p_person)
+
+            bday_index = p_person.birth_ref_index
+            dday_index = p_person.death_ref_index
+
+            if bday_index > -1:
+                bday_handle = p_person.get_event_ref_list()[bday_index].to_struct()['ref']
+                h_event = self.database.get_from_name_and_handle("Event", bday_handle)
+                if h_event.get_type() == "Birth":
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use`int(h_event.get_type()) == Event.BIRTH`.

File: NetworkChart/NetworkChart.py [hunk 10]
```diff
@@ -0,0 +1,1189 @@
...
+                        bday = _('b.unk')
+                if bday == _('b.0000-00-00'):
+                    bday = _('b.unk')
+            else:
+                bday = _('b.unk')
+
+            if dday_index > -1:
+                dday_handle = p_person.get_event_ref_list()[dday_index].to_struct()['ref']
+                h_event = self.database.get_from_name_and_handle("Event", dday_handle)
+                if h_event.get_type() == "Death":
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `int(h_event.get_type()) == Event.DEATH`.

File: NetworkChart/NetworkChart.py [hunk 11]
```diff
@@ -0,0 +1,1189 @@
...
+            else:
+                mother_gref = ''
+
+            # Get edge_marriage
+            h_events = f_family.get_event_list() #get_event_ref_list()
+            if len(h_events) > 0:
+                for handle in h_events:
+                    i_event = self.database.get_from_name_and_handle(
+                        "Event", handle)
+                    if i_event.get_type() == "Marriage":
```
Comment (Nick-Hall) [inline NetworkChart/NetworkChart.py]: Use `int(i_event.get_type()) == Event.MARRIAGE`.

**Comment by zorkmark** [comment]
I'll replace the const_gender with gramps constants, incorporate the
bullets, and update the pull request.

The date displayer will take me a bit longer to investigate but I'll
look into it.

Thank you,

Mark B.##SELECTION_END##

-----Original Message-----

**Comment by sam-m888** [comment]
Menu selection is under Web Reports

network-chart-menu-gramps50

The dialog option screens for the wiki page when ready. 👍 

network-chart-main-options-tab-gramps50
network-chart-colour-options-tab-gramps50
network-chart-privacy-options-tab-gramps50
network-chart-trim-options-tab-gramps50
network-chart-highlight-options-tab-gramps50
network-chart-config-options-tab-gramps50

---
===Usage==
The report is available from the following menu option {{man menu|Reports > Webpages > Network Chart...}}

==Prerequistes==

Requires the following to be installed:
  pip install networkx

To get "libcgraph" for pygraphviz you first need to install the development package of graphviz eg: 
  apt install graphviz-dev

then run:

  pip install pygraphviz

**Comment by zorkmark** [comment]
Thank you for the images. Also, I had missed the dependency on
libcgraph.  

Mark B.

-----Original Message-----

**Comment by sam-m888** [comment]
Thank you for your new report, I've published it.

A few issues so far is the following error is shown in the terminal when the report is initially chosen:
```

(xgrampspy3) [grampsdev@lfs101x-VirtualBox gramps (maintenance/gramps42)]$ ./Gramps.py 
Warning: BooleanOption: Signal recursion blocked. Signals was : value-changed
        from: file: /home/grampsdev/aaSandbox/gramps/gramps/gen/plug/menu/_option.py
              line: 101
              func: set_value
Warning: BooleanOption: Signal recursion blocked. Signals was : value-changed
        from: file: /home/grampsdev/aaSandbox/gramps/gramps/gen/plug/menu/_option.py
              line: 101
              func: set_value
Warning: BooleanOption: Signal recursion blocked. Signals was : value-changed
        from: file: /home/grampsdev/aaSandbox/gramps/gramps/gen/plug/menu/_option.py
              line: 101
              func: set_value
Warning: BooleanOption: Signal recursion blocked. Signals was : value-changed
        from: file: /home/grampsdev/aaSandbox/gramps/gramps/gen/plug/menu/_option.py
              line: 101
              func: set_value
Warning: BooleanOption: Signal recursion blocked. Signals was : value-changed
        from: file: /home/grampsdev/aaSandbox/gramps/gramps/gen/plug/menu/_option.py
              line: 101
              func: set_value
Warning: PersonOption: Exception occurred in callback function.
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gen/utils/callback.py", line 405, in emit
    fn(*args)
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/plug/_guioptions.py", line 711, in __value_changed
    name = _nd.display(self.__db.get_person_from_gramps_id(gid))
  File "/home/grampsdev/aaSandbox/gramps/gramps/gen/display/name.py", line 890, in display
    name = person.get_primary_name()
AttributeError: 'NoneType' object has no attribute 'get_primary_name'
```

The second one is that when you press the **Help**  button it currently goes to:

> https://gramps-project.org/wiki/index.php?title=Gramps_4.2_Wiki_Manual_-_Reports#Network_Chart

instead of:

> https://gramps-project.org/wiki/index.php?title=NetworkChart

**Comment by sam-m888** [comment]
Using [example.gramps](https://gramps-project.org/wiki/index.php?title=Example.gramps) was able to render 2030 individuals using the default settings.

Thank you for these clear charts 👍 

Left hand section of chart [another 10 sections like this 🥇 ]
networkchart-addon-sectionfrom2030individuals-gramps42

**Comment by zorkmark** [comment]
You are welcome.  Also, if you have not already discovered it, in a browser you can hold the cursor over the connector lines and it will give you the GrampsID_from => GrampsID_to.  Sometimes useful when you are zoomed in. 

On the issue of the errors upon initialization, I am seeing something similar.  I will investigate further and provide an update to fix (if I can) and also change the help button to the correct link.  Thank you.

**Comment by sam-m888** [comment]
Thanks for investigating :)

Here's another one maybe related to the issue above:

On "Highlight" tab selecting the [x]Highlight central person graph

and attempting to run the chart give the following error (brings up error
assistant, but does not crash Gramps):


220436: ERROR: _reportdialog.py: line 735: Failed to run report.
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/pygraphviz/agraph.py", line
1566, in __new__
    nh = gv.agnode(graph.handle, n.encode(graph.encoding), _Action.find)
KeyError: 'agnode: no key'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File
"/home/grampsdev/aaSandbox/gramps/gramps/gui/plug/report/_reportdialog.py",
line 699, in report
    MyReport.write_report()
  File
"/home/grampsdev/.gramps/gramps42/plugins/NetworkChart/NetworkChart.py",
line 554, in write_report
    A.get_node(center_person).attr['fillcolor'] = '#FFFD6BFF'
  File "/usr/local/lib/python3.5/dist-packages/pygraphviz/agraph.py", line
424, in get_node
    return Node(self, n)
  File "/usr/local/lib/python3.5/dist-packages/pygraphviz/agraph.py", line
1568, in __new__
    raise KeyError("Node %s not in graph." % n)
KeyError: 'Node  not in graph.'

**Comment by sam-m888** [comment]
On my personal family tree for myself it attempted to generate 10149 nodes

>  Warning: graph has 10149 nodes...layout may take a long time.

(Start: 3.52pm Stopped: 3.59pm) during the run the "Network Chart" dialog
stays on the screen and Gramps greys out in the background with no feed
back on what is happening.

Then an error dialog comes up that stated
..........................................
"Report could not be created - Gramps"
b'out of memory\n'
[Close]
..........................................

Notice the error message has unescaped [b'] and [\n']

Also Unbuntu popped up the "Sorry, the application dot has stopped
unexpectedly"

Gramp did not crash and can be used.

May get a few people unsure as to whether it's working or not, suggest that
the Network Chart's be set by default to show a minimal chart only with
warnings.

**Comment by zorkmark** [comment]
After some investigation, I've discovered that the help_url cannot be set for a REPORT.  The initialization errors are much deeper, although I think that I've figured out why they are occurring.  I will attempt to fix when updating for the Gramps50 version.

In the interim, I did find a couple of minor bugs that I have fixed and will upload as ver_0.0.3.

**Comment by zorkmark** [comment]
The center person was one of the bugs that I have fixed.  

The other issue may be due to the default use of Orthogonal for the Graph Style.  I will modify to default to Curved for databases over 1500 as this has a significant speed up and may solve the time out / crash issue.

**Comment by Paul-Franklin** [comment]
I don't use addons, so I haven't attempted to test yours
(but thanks!), but I have seen lots of bug reports over the
years from users of the GraphView addon.  It is similar to
yours in that it requires the user to have so-and-so also
on their machine.  So while you are still interested in it
I thought I'd ask if there is any possibility (or maybe you
already do it?) that you can add a popup warning dialog
saying something along the lines of "this Network addon
requires you to have so-and-so on your machine and you
don't appear to have it."  Then either tell them how to get
it, or point them at some wiki page which tells them (as
maybe the instructions will be different(?) on Windows or
Mac, than linux?)

But it'd be really nice if we didn't start getting a ton of
bug reports about this one too, simply because the
user doesn't realize they also have to have/get your
library program.  Just a thought.

Thanks.

**Comment by zorkmark** [comment]
The plugin will not load if networkx is not installed.  A log warning message of "Failure to load Family NetworkChart.  Networkx python module not found." is generated.  The module will then not be available for selection in the menus.  This is the current operation.

**Comment by Paul-Franklin** [comment]
Many users do not look at the (printed) log messages,
especially on Windows.  That is why I specifically asked
you to generate a dialog -- in the hope they will see that.

If you are saying that they will get some sort of /easily
noticed/ message from the addon manager -- such as
when they attempt to load /all addons/ -- then I still will
wonder if they will notice whatever it is.

But if you don't want to do anything, that's beyond my
control.  That's why they are "third-party" addons, and
supported by their author.  It's just that (as I have said
elsewhere) that users often don't understand that, and
so file a bug report when it doesn't start.

**Comment by sam-m888** [comment]
> After some investigation, I've discovered that the help_url cannot be set for a REPORT

@zorkmark Yes, that's right forgot about issue [9677: Move "help_url" option to all plugins 'and addons'](https://gramps-project.org/bugs/view.php?id=9677)

**Comment by zorkmark** [comment]
Paul.  Here's the issue.  If you add an ErrorDialog upon failure to load NetworkChart in the gpr.py file as Gramps starts up and there is a failure to load, the ErrorDialog will continue to show up each Gramps restart until the plugin is removed.  However, when I test this the failure to load removes the plugin from the Plugin Manager listing (under the Gramps Help menu) and one would not be able to remove (hide) it from there.  One would have to go into the gramps plugin directory and delete the NetworkChart directory manually to stop the dialog from showing up.  That would surely bring about more bug reports.  

Please explain how to implement such a dialog without this issue and I would be happy to include it.

**Comment by Paul-Franklin** [comment]
Look at the code for the metadata viewer in
plugins/gramplet/gramplet.gpr.py and maybe
that will give you some ideas.  You can click
on that dialog and it will never be shown again.

HTH.

**Comment by uxcq** [comment]
@zorkmark Installed and working 🥇  Excellent report, should be part of Gramps  by default 👍  Thanks for the contribution.

**Comment by Paul-Franklin** [comment]
"Look at the code for the metadata viewer in
plugins/gramplet/gramplet.gpr.py and maybe
that will give you some ideas. You can click
on that dialog and it will never be shown again."

Here is the first of what I hope will /not/ be many more:

https://gramps-project.org/bugs/view.php?id=9986

**Comment by uxcq** [comment]
@Paul-Franklin  Don't think this is fixed by the newer [PR65](https://github.com/gramps-project/addons-source/pull/65) ?    Also a few more mentions over [here](https://www.reddit.com/r/gramps/comments/5z1gtj/show_all_database_entries_in_a_graphical_report/)  and [here](https://www.reddit.com/r/gramps/comments/5zmveg/network_chart_not_showing_in_menu/) as well as people on ( @bpisoj ) Windows and probably MacOS ( @jralls ) wanting to use this addon .

**Comment by Paul-Franklin** [comment]
@uxcq is "Don't think this is fixed by the newer PR65"
a question or a statement?  8-)

**Comment by zorkmark** [comment]
It does not address it.  The metadata viewer gramplet method did not
work, but I have figured out a way that is similar.   Thanks for the
suggestion.   I'll make the addition and upload as ver 0.0.4. 
 ##SELECTION_END##

Also, do I just make a new feature request for bundling the necessary
modules in MacOS and Windows or do I contact the people mentioned
directly?

Thank you,

Mark



-----Original Message-----

**Comment by uli22** [comment]
Hi Devs
I tried to install NetworkChart addon as it was displayed in addon install dialogue but the install showed ERROR without specifying which one. As I use gramps 4.2.5 on Windows it lacks graphviz. there is no way to install development version of pygraphviz on windows.I could install networkx separately.
Any ideas how to run the application?
cheers
Uli

**Comment by jralls** [comment]
You can get a GraphViz windows binary from their website: http://www.graphviz.org/Download_windows.php 
Once you've installed that and added it to the path (see the note at the top of that page) you should be able to install pygraphviz.

**Comment by uxcq** [comment]
> You can get a GraphViz windows binary from their website: http://www.graphviz.org/Download_windows.php

Nope won't work for the Gramps AIO installed version 👎   Needs to be included from the start!

@jralls How does the MacOS version work for addon dependencies? Do they need to be included or can they be added afterwards as I'm seriously considering one of the new [MacBook pro's](http://www.apple.com/au/macbook-pro/) .

**Comment by jralls** [comment]
Josip says that there's already a graphviz in the Windows AIO, so perhaps you can finagle PIP so that the pygraphviz module can find it. You'll want to put pygraphviz and networkx in $HOME/.gramps/gramps42/plugins/NetworkChart (assuming that you're running Gramps 42). Be sure to use the python from the AIO to run PIP!

The Mac AIO also has graphviz included, so it should be the same as Windows except that instead of $HOME/.gramps the plugins folder is at $HOME/Library/Application Support/gramps/gramps42/plugins. It also has its own python, so again be sure to use that and not the system python to run PIP. Some fiddling with PIP

**Comment by uxcq** [comment]
> Josip says that there's already a graphviz in the Windows AIO, so perhaps you can finagle PIP so that the pygraphviz module can find it. You'll want to put pygraphviz and networkx 

@jralls Yes your correct graphviz is already part of Gramps AIO and I was think of pygraphviz. 

Pretty sure you can't run the python in Gramps AIO as it's only a zip file, I'll check if it's possible from the CLI when I wake. 

I'll wait for @bpisoj to put out a new version as he indicates [here](http://gramps.1791082.n4.nabble.com/Add-on-Dependencies-tp4679422p4679437.html)  

> AIO already contain graphviz which is as python himself compiled with
gcc so pygraphviz should be also.
>
> I can try to build it this days!
>
> -- 
> Josip

**Comment by Paul-Franklin** [comment]
The AIO graphviz is called dot.exe and lives in the top folder.

If you cd to that folder (in a command shell) and then say

dot -V

(with a capital-V) it will announce which graphviz version it is.

**Comment by uli22** [comment]
for  me it gives dot - graphviz versioon 2.39.20151222.2251

So if it is allready the right version, will I additionally need pygraphviz? That should be installed allready, too else how can gramps.py use dot.py?
what else is to be done?
cheers 
Uli

**Comment by jralls** [comment]
Gramps itself doesn't use graphviz. A popular plugin, GraphView, does. GraphView interacts with graphviz by forking and execing dot, which is a compiled C  front end to graphviz, instead of interacting with the library via pyGraphviz the way NetworkChart does.

**Comment by uli22** [comment]
So I think I will be a bit more patient and wait for Josip to provide the files. Thank you. 
Cheers
Uli

**Comment by uxcq** [comment]
@bpisoj Where can I find this file?

**Comment by prculley** [comment]
Josip creates new AIO versions of Gramps when we do new releases, (and
sometimes interim releases like for Gramps 5.0).  I expect that unless you
want to set up a full development environment on your windows machine,
including the latest pygraphviz libraries, you will have to wait for a new
release of Gramps.

If you are very adventurous and familiar with software development, you
might want to look at Josip's recent wiki on how to set up the Gramps
portions of a development environment.  You would still be on your own for
some other portions, so don't think of this as a quick way to use this on
Windows.  If you are somewhat proficient with software related things, it
might actually be easier to set up a Linux VM and run the Linux version of
Gramps, where this can be made to work more easily.

https://gramps-project.org/wiki/index.php?title=Gramps_for_Windows_with_MSYS2

Paul Culley

**Comment by uli22** [comment]
I tried networkchart in win 10 with grampsAIO4.2.5 and with development version from gramps master (gramps 5.0).
in both environments I get the following error:
37789: ERROR: _reportdialog.py: line 741: Failed to run report.
Traceback (most recent call last):
  File "C:\msys64\home\Frink\grampsdev\gramps\gui\plug\report\_reportdialog.py", line 702, in report
    my_report.write_report()
  File "C:\Users\Frink\AppData\Roaming/gramps/gramps50/plugins/NetworkChart\NetworkChart.py", line 414, in write_report
    person, edge_marriage, edge_child = self.get_network()
  File "C:\Users\Frink\AppData\Roaming/gramps/gramps50/plugins/NetworkChart\NetworkChart.py", line 239, in get_network
    fmt_bdate = self._datedisplay.display_formatted(h_event.get_date_object())
  File "C:\msys64\home\Frink\grampsdev\gramps\gen\datehandler\_datedisplay.py", line 495, in display_formatted
    text = self.display_cal[cal](start, inflect=date_type)
TypeError: _display_gregorian() got an unexpected keyword argument 'inflect'

which might not be due to the addon but a bug in date displayer.

on gramps 5.0 there are errors in "conv_to_unicode" saying it cannot be found in ...const which can be left out I think
cheers 
Uli

**Comment by zorkmark** [comment]
I haven't run under 5.0 yet, I'll ##SELECTION_END##investigate this
weekend. 

Thank you,

Mark 

-----Original Message-----

**Pull request merged**